### ISSUE TITLE
Add 'yaml-cpp' TPL to Trilinos CI build

### ISSUE THREAD

**Issue by ibaned** [trilinos/Trilinos#1167]
**Next Action Status:**

Waiting for Trilinos to switch CI build from GCC 4.7.2 to GCC 4.7.3 (see #1002) ...

**Description:**

This issue is to get the standard Trilinos CI build to enable the `yaml-cpp` TPL which is now being used by Albany and other customers.

Done for this issue that there exists a Trilinos CI build with the `yaml-cpp` TPL enabled. This TPL already exists as a SEMS module.

**Blocked by:** #1002 (SEMS `yaml-cpp` TPL does not exist for GCC 4.7.2 in SEMS env)

**Comment by bartlettroscoe** [comment]
CC: @trilinos/framework 

It should be noted that Albany has a required dependency on the yaml-cpp TPL and therefore Trilinos needs to add testing of this to its automated tests.  The CI build is just the first built it should be added to.

However, that means that we will need to upgrade Boost used from 1.55.0 to 1.63.0.  See below email (and more info will be added in a later comment).

----

From: Bartlett, Roscoe A 
Sent: Tuesday, March 21, 2017 5:06 PM
To: Trilinos Developers List
Cc: 'Trilinos Framework List'
Subject: Update Trilinos CI build from boost 1.55.0 to 1.63.0?

Hello Trilinos Developers,

I am attempting to add the SEMS-built yaml-cpp 0.5.3 TPL to the set of TPLs tested in the Trilinos CI build.  But this yaml-cpp TPL build requires boost 1.65.0.  The previously defined Boost version for the CI build was 1.55.0.

Does anyone see a problem with upgrading the version of boost from 1.55.0 to 1.63.0 for the Trilinos CI build?

If you have a concern about updating the version of Boost for the standard CI build, please add a comment in:

*  https://github.com/trilinos/Trilinos/issues/1167

Please *don’t* reply this this email.

Thanks,

-Ross

**Comment by mwglass** [comment]
Just a FYI…

I don’t know how much of boost is used by Trilinos but for Sierra, we had to make the following change to boost 1.63 to get things to link properly with the Intel compilers:

diff boost_1_63_0/boost/serialization/array.hpp boost_1_63_0/boost/serialization/array.hpp.fcs
47,48d46
<     // SIERRA: Following ifdef guard necessary for Intel compilers
< #ifndef __INTEL_COMPILER
51d48
< #endif

**Comment by ibaned** [comment]
A couple clarifications:
 - The latest Boost version is 1.63.0, there was a typo about 1.65.0 which isn't out yet.
 - Albany doesn't yet require yaml-cpp, but we have every intention to make it a required dependency as soon as some of our YAML support patches make it into Trilinos master (they are in develop right now, see e.g. a81e0f4).
 - I've built Trilinos against 1.62.0 for some time now with no problems. I'll move to 1.63.0 now.
 - yaml-cpp is actually moving away from Boost, and the current development version has replaced usage of Boost with C++11 constructs (see jbeder/yaml-cpp#342). I don't think a release has been made with since change though.

**Comment by ibaned** [comment]
@mwglass a `grep` of Trilinos at 416fc12 showed no inclusion of `boost/serialization`. I guess its safe from that one issue.

**Comment by mwglass** [comment]
But what about apps that use both boost/serialization and trilinos?

**Comment by ibaned** [comment]
I don't think @bartlettroscoe is suggesting that Trilinos drop support for Boost 1.55.0, only that we use 1.63.0 for the CI build. Applications that support only 1.55.0 but not 1.63.0 can still compile Trilinos against 1.55.0 in their own builds while they work on switching to newer versions.

**Comment by bartlettroscoe** [comment]
> I don't think @bartlettroscoe is suggesting that Trilinos drop support for Boost 1.55.0, only that we use 1.63.0 for the CI build

Correct.  But that is up to the @trilinos/framework team to make sure that happens.

> Applications that support only 1.55.0 but not 1.63.0 can still compile Trilinos against 1.55.0 in their own builds while they work on switching to newer versions.

Yes.

And it was a typo above where I typed 1.65.0 but meant 1.63.0.  I will edit the comment.

**Comment by mwglass** [comment]
OK, thanks for clarifying.

**Comment by jjellio** [comment]
I've been used yaml-cpp from their github site - it does not require boost and I have not had any problems with it so far. I've used it across Cray+Intel, and GCC+Cuda systems so far (KNL, haswell , and GPUs ).

**Comment by bartlettroscoe** [comment]
I was able to pretty quickly add the 'yaml-cpp' TPL to the Trilinos SEMS Dev Env (see details below).  I pushed this update to the branch:

* https://github.com/bartlettroscoe/Trilinos/tree/add-yaml-cpp-to-ci-build-1167

I tested that it is enabling and passing the `TeuchosParameterList_YamlParser_test_MPI_1` test for both the GCC 4.8.4 and GCC 4.9.3 SEMS Dev Env.

Unfortunately, I can't push this to the Trilinos 'develop' branch because SEMS does not build the yaml-cpp TPL for GCC 4.7.2, which is used by the current CI build.  I think this means that we will not be able to add yaml-cpp to the standard CI build until we upgrade it from GCC 4.7.2 to 4.9.3 (see #1002).  (There is a very important Trilinos customer that still requires GCC 4.7.2 support so we can't upgrade from GCC 4.7.2 right now unfortunately.)

Therefore, we need to block this issue with #1002.  As soon as the standard CI build is using GCC 4.9.3, then we just need to merge and push this branch and this will be done.

----

**DETAILED NOTES:**

First I need to update the SEMS Dev Env on my machine crf450 (which syncs it to speed up usage).  For that, I just need to run:

```
$ cd
$ time ./update_local_sems_env_sync.sh &> update_local_sems_env_sync.out

real    40m25.755s
user    4m12.009s
sys     5m29.637s
```

However, I don't see that the yaml-cpp TPL was copied over:

```
$ grep yaml update_local_sems_env_sync.out
[empty]
```

Darn, my script update_local_sems_env_sync.sh was incorrect.  It was not correctly pulling from the repo sems-users.  I fixed that now am running again ...

```
$ cd
$ time ./update_local_sems_env_sync.sh &> update_local_sems_env_sync.out

real    16m5.219s
user    1m38.399s
sys     2m14.962s
```

Now the yaml-cpp SEMS module is shown (using tab completion with `module load sems-y [tab]`):

```
$ module load sems-yaml_cpp/0.5.3/base
```

Now to add yaml-cpp to the Standard Trilinos SEMS Dev Env and CI build:

**1) Add yaml-cpp module info to Trilinos/cmake/std/sems/get_default/modules.sh and Trilinos/cmake/load_sems_dev_env.sh:**

I made that obvious addition for loading the module `sems-yaml_cpp/0.5.3/base` but I got the following warning:

```
$ . ~/Trilinos.base/Trilinos/cmake/load_sems_dev_env.sh 
        WARNING: sems-boost dependency already found but does not match listed dependency sems-boost/1.63.0/base
        I will use the sems-boost you have loaded but correct behavior is not guaranteed
```

I appears that the SEMS-built `sems-yaml_cpp/0.5.3/base` module depends on the `sems-boost/1.63.0/base` TPL, not the `sems-boost/1.55.0` TPL that was being loaded by the `load_sems_eve_env.sh` source script.

Therefore, I updated the Boost version used from 1.55.0 to 1.63.0 to be consistent.  I sent the Trilinos developers an email asking if this was a problem or not.

We have another problem.  The standard CI build uses GCC 4.7.2 and the yaml-cpp TPL is not provided by SEMS for that compiler as shown by:

```
. ~/Trilinos.base/Trilinos/cmake/load_ci_sems_dev_env.sh 
        ERROR loading module sems-yaml_cpp/0.5.3/base
        Module points to path  /projects/sems/install/rhel6-x86_64/sems/tpl/yaml_cpp/0.5.3/gcc/4.7.2/base  which does not exist
        You may need to load the appropriate compilers before loading this module
```

I think this means that we will not be able to add yaml-cpp to the standard CI build until we upgrade it from GCC 4.7.2 to 4.9.3 (see #1002).  However, I can still build with GCC 4.8.4 which is the default compiler loaded by the Trilinos/cmake/load_sems_dev_env.sh script.

**2) Add find info for yaml-cpp to the file Trilinos/cmake/std/sems/SEMSDevEnv.cmake:**

Just a simple block of CMake code.

**3) Add default enable for yaml-cpp to Trilinos/cmake/std/BasicCiTestingSettings.cmake:**

I added the enable.

I pushed these changes to the branch:

```
To git@github.com:bartlettroscoe/Trilinos
 * [new branch]      add-yaml-cpp-to-ci-build-1167 -> add-yaml-cpp-to-ci-build-1167
```

**4) Testing with GCC 4.8.4 SEMS Env:**

Using the `do-configure` script:

```
#!/bin/bash

cmake \
-DTrilinos_TRIBITS_DIR:STRING=TriBITS/tribits \
-DTPL_ENABLE_MPI=ON \
-DCMAKE_BUILD_TYPE=RELEASE \
-DTrilinos_ENABLE_DEBUG=ON \
-DBUILD_SHARED_LIBS=ON \
-DTrilinos_ENABLE_SECONDARY_TESTED_CODE=OFF \
-DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake \
-DTrilinos_ENABLE_DEBUG_SYMBOLS=ON \
-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION=ON \
-DTrilinos_ENABLE_CONFIGURE_TIMING=ON \
-DTrilinos_ENABLE_TESTS=ON \
-DTrilinos_TRACE_ADD_TEST=ON \
-DCTEST_BUILD_FLAGS=-j16 \
-DCTEST_PARALLEL_LEVEL=16 \
"$@" \
../../../Trilinos
```

Testing Teuchos with yaml-cpp enabled on my machine crf450 with:

```
$ cd /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.4/MPI_RELEASE_DEBUG_SHARED_PT/

$ rm -r CMake*

$ . ~/Trilinos.base/Trilinos/cmake/load_sems_dev_env.sh 

$ which gcc
/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.8.4/base/bin/gcc

$ module list
Currently Loaded Modulefiles:
  1) sems-env                       4) sems-git/2.10.1                7) sems-boost/1.63.0/base        10) sems-hdf5/1.8.12/parallel     13) sems-superlu/4.3/base
  2) sems-python/2.7.9              5) sems-gcc/4.8.4                 8) sems-yaml_cpp/0.5.3/base      11) sems-netcdf/4.3.2/parallel
  3) sems-cmake/3.5.2               6) sems-openmpi/1.8.7             9) sems-zlib/1.2.8/base          12) sems-parmetis/4.0.3/parallel

$ time ./do-configure -DTrilinos_ENABLE_Teuchos=ON &> configure.out 

real    0m10.764s
user    0m6.412s
sys     0m2.683s

$ time make -j32 &> make.out

real    2m17.493s
user    24m56.503s
sys     4m10.020s

$ time ctest -j32 &> ctest.out

real    0m2.937s
user    0m7.743s
sys     0m15.388s
```

This returned all passing tests:

```
100% tests passed, 0 tests failed out of 125

Label Time Summary:
Teuchos    =  62.88 sec (125 tests)

Total Test time (real) =   2.93 sec
````

And we see YAML tests as well:

```
$ grep -i yaml ctest.out 
        Start  80: TeuchosParameterList_YamlParser_test_MPI_1
 55/125 Test  #80: TeuchosParameterList_YamlParser_test_MPI_1 ......................................   Passed    0.54 sec
```

You can also use the base checkin-test.py script by loading the GCC 4.8.4 SEMS Dev Env as:

```
$ mkdir CHECKIN_GCC-4.8.4/
$ cd CHECKIN_GCC-4.8.4/
$ ln -s ../../Trilinos/checkin-test.py .
$ ./checkin-test.py -j16 --enable-all-packages=off --no-enable-fwd-packages \
  --enable-packages=Teuchos --local-do-all
```

The resulted in:

```
0) MPI_RELEASE_DEBUG_SHARED_PT Results:
---------------------------------------

  passed: Trilinos/MPI_RELEASE_DEBUG_SHARED_PT: passed=125,notpassed=0
  
  Tue Mar 21 16:26:43 MDT 2017
  
  Enabled Packages: Teuchos
  Disabled Packages: PyTrilinos,Claps,TriKota
  Hostname: crf450.srn.sandia.gov
  Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
  Build Dir: /home/rabartl/Trilinos.base/BUILDS/CHECKIN_GCC-4.8.4/MPI_RELEASE_DEBUG_SHARED_PT
  
  CMake Cache Varibles: -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=BASIC -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=300.0 -DBUILD_SHARED_LIBS=ON -DTrilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=ON -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_SECONDARY_TESTED_CODE:BOOL=OFF -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake -DTrilinos_ENABLE_Teuchos:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=OFF -DTrilinos_ENABLE_PyTrilinos:BOOL=OFF -DTrilinos_ENABLE_Claps:BOOL=OFF -DTrilinos_ENABLE_TriKota:BOOL=OFF
  Make Options: -j16 
  CTest Options: -j16 
  
  Pull: Not Performed
  Configure: Passed (0.15 min)
  Build: Passed (2.13 min)
  Test: Passed (0.06 min)
  
  100% tests passed, 0 tests failed out of 125
  
  Label Time Summary:
  Teuchos    =  35.06 sec (125 tests)
  
  Total Test time (real) =   3.56 sec
  
  Total time for MPI_RELEASE_DEBUG_SHARED_PT = 2.34 min
```

So if the CI build usesd GCC 4.8.4, we would be all good.

Also trying this for GCC 4.9.3:

```
$ cd CHECKIN_GCC-4.9.3/

$ . ~/Trilinos.base/Trilinos/cmake/load_sems_dev_env.sh sems-gcc/4.9.3

$ module list
Currently Loaded Modulefiles:
  1) sems-env                       4) sems-git/2.10.1                7) sems-boost/1.63.0/base        10) sems-hdf5/1.8.12/parallel     13) sems-superlu/4.3/base
  2) sems-python/2.7.9              5) sems-gcc/4.9.3                 8) sems-yaml_cpp/0.5.3/base      11) sems-netcdf/4.3.2/parallel
  3) sems-cmake/3.5.2               6) sems-openmpi/1.8.7             9) sems-zlib/1.2.8/base          12) sems-parmetis/4.0.3/parallel

$ ./checkin-test.py --enable-all-packages=off --no-enable-fwd-packages \
  --enable-packages=Teuchos --local-do-all
```

That returned:

```
0) MPI_RELEASE_DEBUG_SHARED_PT Results:
---------------------------------------

  FAILED: Trilinos/MPI_RELEASE_DEBUG_SHARED_PT: passed=124,notpassed=1
  
  Tue Mar 21 16:35:50 MDT 2017
  
  Enabled Packages: Teuchos
  Disabled Packages: PyTrilinos,Claps,TriKota
  Hostname: crf450.srn.sandia.gov
  Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
  Build Dir: /home/rabartl/Trilinos.base/BUILDS/CHECKIN_GCC-4.9.3/MPI_RELEASE_DEBUG_SHARED_PT
  
  CMake Cache Varibles: -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=BASIC -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=300.0 -DBUILD_SHARED_LIBS=ON -DTrilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=ON -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_SECONDARY_TESTED_CODE:BOOL=OFF -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake -DTrilinos_ENABLE_Teuchos:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=OFF -DTrilinos_ENABLE_PyTrilinos:BOOL=OFF -DTrilinos_ENABLE_Claps:BOOL=OFF -DTrilinos_ENABLE_TriKota:BOOL=OFF
  Make Options: -j16 
  CTest Options: -j16 
  
  Pull: Not Performed
  Configure: Passed (0.21 min)
  Build: Passed (2.94 min)
  Test: FAILED (0.06 min)
  
  99% tests passed, 1 tests failed out of 125
  
  Label Time Summary:
  Teuchos    =  37.35 sec (125 tests)
  
  Total Test time (real) =   3.80 sec
  
  The following tests FAILED:
         66 - TeuchosParameterList_ObjectBuilder_UnitTests_MPI_1 (Failed)
  Errors while running CTest
  
  Total time for MPI_RELEASE_DEBUG_SHARED_PT = 3.22 min
```

Darn, we get a test failure (that test is problematic anyway).  It segfaulted.  But it did pass the YAML test:

```
$ grep -i yaml MPI_RELEASE_DEBUG_SHARED_PT/ctest.out 
        Start  80: TeuchosParameterList_YamlParser_test_MPI_1
 77/125 Test  #80: TeuchosParameterList_YamlParser_test_MPI_1 ......................................   Passed    0.33 sec
```

So, more or less, all is good with GCC 4.9.3 as well (at least for yaml-cpp).

All is well except we can't use GCC 4.7.2 so we can't add yaml-cpp to the standard CI build.  Therefore, I can't push this to the 'develop' branch or it would break the CI build.

**Comment by bartlettroscoe** [comment]
I created the SEMS service request:

* https://sems.sandia.gov/jira/servicedesk/customer/portal/4/SEMSHELPD-408

 provide the yaml-cpp TPL for GCC 4.7.2.

From looking at:

* https://sems.sandia.gov/jira/browse/SEMS-1259

it appears that yaml-cpp was installed to support Nalu (and important internal Trilinos customer).

But there is no indication why yaml-cpp was not provided for GCC 4.7.2 or if anyone considered Trilinos specifically.  It appears that other details and clues might be in SEMSHELPD-66 but no one can see that except for the direct customer and SEMS staff.

I am guessing that Albany developers knew nothing about this SEMS effort to provide yaml-cpp?

**Comment by nmhamster** [comment]
@bartlettroscoe I agree with @jjellio that testbeds is using `yaml-cpp` without Boost. I would try to avoid this being needed if we can right?

**Comment by ibaned** [comment]
> I am guessing that Albany developers knew nothing about this SEMS effort to provide yaml-cpp?

As far as I know thats correct. We were simply watching its growing support in Teuchos and MueLu. I am thankful that Nalu got things going with SEMS, and I hope the timeframe for resolving #1002 is not too far away. For Albany though, our only real roadblock is the merging of Trilinos develop into master.

**Comment by bartlettroscoe** [comment]
>  I agree with @jjellio that testbeds is using yaml-cpp without Boost. I would try to avoid this being needed if we can right?

Then perhaps we can get SEMS to build the version of yaml-cpp that does not require Boost.  But does that version work with GCC 4.7.2?  It would be good of the SEMS env and the test beds would use consistent versions of the TPLs.

I created the SEMS service request:

* https://sems.sandia.gov/jira/servicedesk/customer/portal/4/SEMSHELPD-409

to use the version of yaml-cpp that does not depend on boost.  But does that version build with C++11 support in GCC 4.7.2?  If not, this will still have to wait for #1002.

Boy, it seems like what should be the simplest things almost always turns out to be hard or impossible when dealing with the confluence of Trilinos, SEMS, and the various Trilinos customers.  There are just so many constraints.

**Comment by lxmota** [comment]
I agree with @ibaned that for Albany the only roadblock right now is the merging of Trilinos develop into master for this YAML issue and another issue as well (#1132).

**Comment by nmhamster** [comment]
> Then perhaps we can get SEMS to build the version of yaml-cpp that does not require Boost. But does that version work with GCC 4.7.2? It would be good of the SEMS env and the test beds would use consistent versions of the TPLs.

@bartlettroscoe we are working towards this goal but not all packages and compilers work on all platforms (for instance, there is no GCC 4.7 on POWER with Little Endian). 

> Boy, it seems like what should be the simplest things almost always turns out to be hard or impossible when dealing with the confluence of Trilinos, SEMS, and the various Trilinos customers. There are just so many constraints.

Yes. That's possibly because we are not aggressively pushing updates into the code to keep up with newer TPLs? Perhaps this is something with a little focus we can improve on, having all the newer TPLs available for testing is a good first step though.

**Comment by bartlettroscoe** [comment]
Now that we are pulling the trigger on the upgrade to GCC 4.8.3 in #1002, I am going to try enabling yaml-cpp as well.  But note, as mentioned [above](https://github.com/trilinos/Trilinos/issues/1167#issuecomment-288218664), this will require upgrading Boost from 1.55.0 to 1.63.0.  If there is an objection to this let me know.  But with SIERRA adding their own specialized testing for Trilinos, I don't know that it is important for the CI build of Trilinos to match the version of boost used by SIERRA.  Or we could ask SEMS to install a different yaml-cpp version that does not depend on Boost?

**Comment by bartlettroscoe** [comment]
Now that we are pulling the trigger on the upgrade to GCC 4.8.3 in #1002, I am going to try enabling yaml-cpp as well.  But note, as mentioned [above](https://github.com/trilinos/Trilinos/issues/1167#issuecomment-288218664), this will require upgrading Boost from 1.55.0 to 1.63.0.  If there is an objection to this let me know.  But with SIERRA adding their own specialized testing for Trilinos, I don't know that it is important for the CI build of Trilinos to match the version of boost used by SIERRA.  Or we could ask SEMS to install a different yaml-cpp version that does not depend on Boost?

**Comment by bartlettroscoe** [comment]
I am trying this out right now as part of #1002.

**Comment by csiefer2** [comment]
There is no sems gcc 4.7.2 yaml-cpp TPL.

This is no bueno.

@maherou  @bmpersc

**Comment by bartlettroscoe** [comment]
@csiefer2:

> There is no sems gcc 4.7.2 yaml-cpp TPL.

That was a know issue (see [above](https://github.com/trilinos/Trilinos/issues/1167#issuecomment-288267046)).  This is no problem now that the CI build has been updated to use GCC 4.9.3 (see #1002).  Let me know if you have a problem with this.

**Comment by bartlettroscoe** [comment]
The yaml-cpp is not a PT TPL and include in standard CI testing now that GCC 4.9.3 is the selected compiler for the CI build (see #1002) the commit  e0d2865 has been pushed to 'develop'.  We can see the Teuchos YAML reader being tested in the current post-push CI build running at:

* http://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=2890007&filtercount=3&showfilters=1&field1=groupname&compare1=61&value1=Continuous&field2=buildstarttime&compare2=84&value2=now&filtercombine=and

and you can see the test TeuchosParameterList_YamlParser_test_MPI_1 being run and passing in:

* http://testing.sandia.gov/cdash/viewTest.php?onlypassed&buildid=2890011

Closing as complete.  I don't see any need to leave this in review (the review evidence is given above).
